src/net/tcp/nbd.c

```diff
@@ -0,0 +1,1095 @@
+ * @v buffer		Data buffer
+ * @v len		Length of data buffer
+ * @ret rc		Return status code
+ */
+static int nbd_command ( struct nbd_session *nbd, struct interface *block,
+			 enum nbd_command_type type,
+			 uint64_t lba, unsigned int count,
+			 userptr_t buffer, size_t len ) {
+	struct nbd_command *nbdcmd;
+
+	/* Sanity check */
+	assert ( nbd->command == NULL && nbd->rx_state == NBD_RX_TRANS_REP_CMD );
+
+	/* Allocate and initialise structure */
+	nbdcmd = zalloc ( sizeof ( *nbdcmd ) );
+	if ( ! nbdcmd )
+		return -ENOMEM;
+
+	ref_init ( &nbdcmd->refcnt, nbdcmd_free );
+	intf_init ( &nbdcmd->block, &nbdcmd_block_desc, &nbdcmd->refcnt );
+
+	nbdcmd->nbd = nbd;
+	nbd->command = nbdcmd;
+	ref_get ( &nbdcmd->nbd->refcnt );
+
+	nbdcmd->type = type;
+	nbdcmd->lba = lba;
+	nbdcmd->count = count;
+
+	nbdcmd->data_buffer = buffer;
+	nbdcmd->data_len = len;
+
+	/* Attach to parent interface, transfer reference to session,
+	 * and return
+	 */
+	intf_plug_plug ( &nbdcmd->block, block );
+	return 0;
+}
+
+
+/**
+ * Free NBD session
+ *
+ * @v refcnt		Reference counter
+ */
+static void nbd_free ( struct refcnt *refcnt ) {
+	struct nbd_session *nbd =
+		container_of ( refcnt, struct nbd_session, refcnt );
+
+	uri_put ( nbd->uri );
+	free ( nbd );
+}
+
+/**
+ * Shut down NBD interface
+ *
+ * @v nbd		NBD session
+ * @v rc		Reason for close
+ */
+static void nbd_close ( struct nbd_session *nbd, int rc ) {
+	/* A TCP graceful close is still an error from our point of view */
+	if ( rc == 0 )
+		rc = -ECONNRESET;
+
+	DBGC ( nbd, "NBD %p closed: %s\n", nbd, strerror ( rc ) );
+
+	/* Stop transmission process */
+	process_del ( &nbd->process );
+
+	if ( nbd->command )
+		nbdcmd_close ( nbd->command, rc );
+
+	/* Shut down interfaces */
+	intfs_shutdown ( rc, &nbd->block, &nbd->socket, NULL );
+}
+
+
+/****************************************************************************
+ *
+ * Block to NBD interface
+ *
+ */
+
+/**
+ * Check NBD flow-control window
+ *
+ * @v nbd		NBD session
+ * @ret len		Length of window
+ */
+static size_t nbd_block_window ( struct nbd_session *nbd ) {
+	DBGC ( nbd, "NBD %p block window\n", nbd );
+
+	if ( nbd->rx_state >= NBD_RX_TRANS_REP_CMD &&
+	     nbd->command == NULL )
+		/* We cannot handle concurrent commands */
+		return 1;
+	else
+		return 0;
+}
+
+/**
+ * Issue NBD block read
+ *
+ * @v nbd		NBD session
+ * @v block		Block data interface
+ * @v lba		Starting logical block address
+ * @v count		Number of blocks to transfer
+ * @v buffer		Data buffer
+ * @v len		Length of data buffer
+ * @ret rc		Return status code
+
+ */
+static int nbd_block_read ( struct nbd_session *nbd,
+			    struct interface *block,
+			    uint64_t lba, unsigned int count,
+			    userptr_t buffer, size_t len ) {
+	int rc;
+
+	DBGCP ( nbd, "NBD %p block %p read LBA 0x%08llx count 0x%04x\n",
+		nbd, block, lba, count );
+
+	rc = nbd_command ( nbd, block, NBD_BLOCK_READ,
+			   lba, count, buffer, len );
+	if ( rc != 0 )
+		return rc;
```

Comment:
Please use the more iPXE-idiomatic:
```
if ( ( rc = nbd_command (......) ) != 0 )
    return rc;
```
here and at all other locations within this code

Reply:
OK, fixed